src/main/java/br/ufpe/cin/files/FilesManager.java

```diff
@@ -220,17 +230,77 @@ private FilesManager(){}
 	 * @return string content of the file, or null in case of errors.
 	 */
 	public static String readFileContent(File file){
+		
 		//StringBuilder content = new StringBuilder();
 		String content = "";
 		try{
-			BufferedReader reader = Files.newBufferedReader(Paths.get(file.getAbsolutePath()), StandardCharsets.UTF_8);
+			String fileEncoding = encodings.get(file);
+			BufferedReader reader = Files.newBufferedReader(Paths.get(file.getAbsolutePath()), Charset.forName(fileEncoding));
 			content = reader.lines().collect(Collectors.joining("\n"));
 		}catch(Exception e){
 			//System.err.println(e.getMessage());
 		}
 		return content;
 	}
-
+	
+	/**
+	 * Try to detect the encoding of the files envolved in a merge.
+	 * @param files to be analysed
+	 */
+	public static void detectEncoding(File left, File base, File right, boolean detectionEnabled) {
+		
+		encodings = new HashMap<File, String>();
+		
+		try {
+			
+			if(detectionEnabled) {	
+				detectEncoding(left);
+				detectEncoding(base);
+				detectEncoding(right);
+			} 
+			
+			else {
+				encodings.put(left, "UTF-8");
+				encodings.put(base, "UTF-8");
+				encodings.put(right, "UTF-8");
+			}
+				
+			} catch (IOException e) {
+				System.err.println("An error occurred while opening files for encoding detection.");
+				System.exit(-1);
+			}
+	}
+	
+	/**
+	 * Try to detect the encoding of a file using Mozilla's API.
+	 * @param file to be analysed
+	 * @throws IOException 
+	 */
+	private static void detectEncoding(File file) throws IOException {
+		
+		if(file != null) {
```

Comment:
Faz sentido o arquivo ser nulo?

Reply:
Acho melhor... Até porque o ideal é quem chama o método ter essa preocupação. Além disso, quando o arquivo é nulo, o método termina não fazendo nada silenciosamente, não deixando isso claro.

Reply:
Vejam o JavaDoc, pessoal, pra entender melhor os por menores da ferramenta. Um arquivo pode ser null sim:

![image](https://user-images.githubusercontent.com/3333981/45522916-79243f80-b79c-11e8-85f0-78d629cb95e6.png)

Reply:
De qualquer forma, existem várias verificações de arquivo nulo no código, então talvez seja algo proposital.

Reply:
@guilhermejccavalcanti certo, mas é que a ferramenta não deixa você chamar arquivos nulos pela linha de comando. Só por código, certo?

Reply:
Acho que a questão é que o arquivo pode ser nulo/vazio quando foi removido em uma das partes.

Reply:
Mesmo por linha de comando deveria ser possível. Null é usado para quando alguma das versões do arquivo foi apagada. Isso é mais comum em execuções de experimentos, através do método ```mergeRevisions``` , que no fim das contas invoca o ```mergeFiles```. No uso real, com o git, esse null não acontece porque o git trata isso antes de invocar a ferramenta.

Imagino que essas dúvidas se devem ao método ```FilesManager.validateFiles``` logo no começo, né? Mas observa direitinho que a condição que tem lá é ``` if(f!=null && !f.exists())```

Sobre o teste que tu falou, tá barrando isso, é? Qual é o teste?

Reply:
> Não, não faz, mas a nível de código, é possível chamar um merge com um arquivo nulo (há um teste que faz isso). Isso é impraticável ao executar a ferramenta, porém. Para não travar o teste, eu coloquei esse if temporariamente, mas posso alterar o teste também.

@jvcoutinho  não sei se entendi bem, mas você não fez isso para o teste passar. fez isso porque é o comportamento esperado da ferramenta.

Reply:
> Acho melhor... Até porque o ideal é quem chama o método ter essa preocupação. Além disso, quando o arquivo é nulo, o método termina não fazendo nada silenciosamente, não deixando isso claro.

@gaabs não acho que o ideal é quem chama ter a preocupação. onde você viu essa recomendação? isso pode até favorecer performance (nesse caso provavelmente a diferença não seria significativa, o ganho sendo irrelevante), mas compromete corretude, principalmente se há o potencial desse método ser usado em outros contextos. por isso que eu prefiro "quem oferece a funcionalidade é que deveria ter o cuidado para ela ser oferecida de forma segura". agora, claro, o método deveria sinalizar o problema (não terminar silenciosamente)! @jvcoutinho @guilhermejccavalcanti isso de terminar silenciosamente é muito ruim. mudem isso sempre que forem vendo algo nessa linha

Reply:
@pauloborba Eu estava assumindo que era esperado que o arquivo nunca fosse nulo. Daí, não faria sentido repassar isso adiante.

Além disso, Bob Martin também recomenda não repassar nulo adiante no Clean Code, no capítulo 7: "In most programming languages there is no good way to deal with a null that is passed by a caller accidentally. Because this is the case, the rational approach is to forbid passing null by default. When you do, you can code with the knowledge that a null in an argument list is an indication of a problem, and end up with far fewer careless mistakes.".

Também li um post recentemente, falando nesse sentido: "It is much better as a software shop to put the onus of not passing null on the caller of any method rather than putting it on the writer of the method being called." (https://elegantcode.com/2010/05/01/say-no-to-null/)

Reply:
@gaabs não repassar null, tudo bem, não tenho nada contra. mas é difícil garantir isso, principalmente se seu código vai ser usado por terceiros. muitas vezes não é realista também, como em m(metododeapideterceirosquepoderetornarnull())... daí a necessidade de tratar o null no método que você está escrevendo; pelo menos com um assert.